### ISSUE TITLE
Always dependencies unmet while compiling in Ubuntu 16.04

### ISSUE THREAD

**Issue by muronghan** [shadowsocks/shadowsocks-libev#1274]
### What version of shadowsocks-libev are you using?
3.0.2  2e2f839

### What operating system are you using?
Ubuntu 16.04

### What did you do?
Compile ss directly from git source

### What did you expect to see?
Compiled successfully

### What did you see instead?
Tried different ways but failed to compile it

### What is your config in detail (with all sensitive info masked)?
n/a

I guess the readme misled.:
- e.g. pkg-config may have to be installed but it was not included in the command line, 
- e.g. it might be "dpkg-buildpackage -b -us -uc -d" while a "-d" was missing otherwise it said "Unmet build dependencies: debhelper (>= 10) libsodium-dev (>= 1.0.8~)"
- Finally it said:" dpkg-shlibdeps: error: no dependency information found for /usr/lib/libmbedcrypto.so.0 (used by debian/shadowsocks-libev/usr/bin/ss-tunnel)" but that file was there under /usr/lib as I maually installed libmbedtls.

Tired many ways but failed. Meantime, ./scripts/deb4xenial.sh did not work either.

**Comment by rogers0** [comment]
> Meantime, ./scripts/deb4xenial.sh did not work either.

Please report what's your problem.

**Comment by maddie** [comment]
Same issue here. Log: http://pastebin.com/jC6ez9Fu

Building with `./script/deb4xenial.sh` under 32-bit Ubuntu 16.04.2 with all updates installed.

**Comment by henrypijames** [comment]
Did you perhaps not install libmbedtls and libsodium, as stated in Readme section Prerequisites?

**Comment by rogers0** [comment]
@maddie "gbp:error: Can't run inside a git repository."

Please retry in a clean place.

**Comment by maddie** [comment]
@rogers0 it's actually being run in a fresh cloned repository.

**Comment by rogers0** [comment]
@maddie try commands below:
```shell
cp script/deb4xenial.sh /dev/shm/
cd $_
./deb4xenial.sh
```

**Comment by rogers0** [comment]
@maddie if you still meet issue, please try install debhelper at the begining
```shell
sudo apt-get install -y -t xenial-backports debhelper
cp script/deb4xenial.sh /dev/shm/
cd $_
./deb4xenial.sh
```

**Comment by asmadeus08** [comment]
wget --no-check-certificate https://raw.githubusercontent.com/teddysun/shadowsocks_install/master/shadowsocks-libev-debian.sh
chmod +x shadowsocks-libev-debian.sh
./shadowsocks-libev-debian.sh 2>&1 | tee shadowsocks-libev-debian.log

**Comment by muronghan** [comment]
@asmadeus08 it worked on my Debian 8 (just upgrade jessie from wheezy and run the .sh).

**Comment by rogers0** [comment]
try **build_deb.sh** script described in https://github.com/shadowsocks/shadowsocks-libev#build-deb-package-from-source

**Comment by muronghan** [comment]
@rogers0 Failed to do the new build_deb.sh script on my Ubuntu 14.04. But that's a 2.6.x kernel. Today probably will try another 14.04 (kernel 3.13.x) one and let you know.

If using the script, it is not necessary to manually install libsodium and libmbedtls, right?

**Comment by debiansid** [comment]
Cool ,let me try it on 14.04

**Comment by maddie** [comment]
With the `build_deb.sh` script, now builds successfully.

**Comment by muronghan** [comment]
build_deb.sh still failed on my fresh Ubuntu 14.04 (Kernel 3.13.x):

There is something wrong as below I guess:

.......
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"
...
**gbp:error: Git command failed: Error running git clone: fatal: destination path 'shadowsocks-libev' already exists and is not an empty directory.**
...
Setting up shadowsocks-libev-build-deps (3.0.2+ds-1~exp2) ...
Processing triggers for libc-bin (2.19-0ubuntu6.9) ...

****gbp:error: You have uncommitted changes in your source tree:
gbp:error: On branch master
Your branch is up-to-date with 'origin/master'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        modified:   debian/compat
        modified:   debian/control
        modified:   debian/rules**


gbp:error: Use --git-ignore-new to ignore.**

**Comment by rogers0** [comment]
@muronghan nice catch!
but I think it's easy for you to fix this.
kinda minor issue, I'll update script next time. thanks!

**Comment by rogers0** [comment]
@muronghan sorry, it's not the problem I was thinking.

obviously, your system is not fresh enough:
fatal: destination path 'shadowsocks-libev' already exists and is not an empty directory.

**Comment by muronghan** [comment]
@rogers0 , below was what I did:

step 1: rebuild Ubuntu Server 16.04 on a VPS;
step 2: apt update && apt dist-upgrade;
step 3: apt install git;
step 4:
git clone https://github.com/shadowsocks/shadowsocks-libev.git
cd shadowsocks-libev
git submodule update --init --recursive
step 5: run your script

... I guess I should skip step 4? I will try it tonight. If yes, I guess the readme should be modified a little bit to be more precise.

**Comment by maddie** [comment]
@muronghan I think the problem is the script should be ran in a clean directory, which is stated in the README:

```
mkdir -p ~/build-area/
cp ./scripts/build_deb.sh ~/build-area/
cd ~/build-area
./build_deb.sh
```

As you can see, the script is being copied into a new directory `~/build-area` and executed inside it.

**Comment by debiansid** [comment]
why you build it on your vps?

**Comment by muronghan** [comment]
@maddie actually I did reboot VPS, re-login and "cd /tmp" before git clone the shadowsocks-libev sources, so that the whole shadowsocks-libev directory was under /tmp/. Supposedly it was a pure clean installation?

Well I think step 4 can not be skipped otherwise the ./scripts/build_deb.sh would not be copied and executed.

So @rogers0 you should check it again?

**Comment by muronghan** [comment]
@debiansid where did you build your .deb? On a virtual machine? Suppose if you have 3 VPSs with different structure e.g. KVM, XEN, OpenVZ and installed with different OS e.g. Debian 8.0, Ubuntu 16.04, Ubuntu 14.04 etc. Can a same .deb work for all platform?

**Comment by maddie** [comment]
@muronghan Assuming `shadowsocks-libev` is cloned under your home directory, can you try:

```
mkdir -p ~/build-area/
cp ~/shadowsocks-libev/scripts/build_deb.sh ~/build-area/
cd ~/build-area
./build_deb.sh
```

I have successfully built debs using the exact commands under Ubuntu 16.04 32-bit & 64-bit

**Comment by debiansid** [comment]
@muronghan 
I compile .deb local on my ubuntu 14.04 build machine,then upload to my VPS. so I don't need building environment on any VPS.

**Comment by muronghan** [comment]
@maddie OK, I will try to git clone to my home directory instead of /tmp/ and then take a shot tonight.

**Comment by debiansid** [comment]
@muronghan you should keep all VPSs in same version of OS, it will make your life easier.

**Comment by muronghan** [comment]
@debiansid you might be right. However, the risk is there if you always choose to upgrade your ss to the newest version. And then you wife will drive you made, haha...

**Comment by debiansid** [comment]
@muronghan
As long as showsocks-libev is written in C, you can always upgrade it in old system.

**Comment by muronghan** [comment]
Ok, @rogers0  I show you the most interesting part of my test:

I did fresh rebuild for Ubuntu 14.04x64 and 16.04x64 respectively. apt update and upgrade. apt install git. And then git clone shadowsocks-libev source to my home directory (I guest that was /root as I got logged in as root).

And then follow your .sh instruction. The compiling results on 14.04 and 16.04 were completely different. It failed on 14.04 and successfully done on 16.04. 

Hope that will help you for further debug....OK I will take another shot on Debian 8...
